producer/src/main/java/com/flipkart/varadhi/produce/services/ProducerService.java

```diff
@@ -1,134 +1,267 @@
 package com.flipkart.varadhi.produce.services;
 
+import com.flipkart.varadhi.common.EntityReadCache;
+import com.flipkart.varadhi.common.EntityReadCacheRegistry;
 import com.flipkart.varadhi.common.Result;
-import com.flipkart.varadhi.common.VaradhiCache;
-import com.flipkart.varadhi.entities.*;
 import com.flipkart.varadhi.common.exceptions.ProduceException;
 import com.flipkart.varadhi.common.exceptions.ResourceNotFoundException;
-import com.flipkart.varadhi.common.exceptions.VaradhiException;
+import com.flipkart.varadhi.entities.InternalCompositeTopic;
+import com.flipkart.varadhi.entities.Message;
+import com.flipkart.varadhi.entities.Offset;
+import com.flipkart.varadhi.entities.StorageTopic;
+import com.flipkart.varadhi.entities.VaradhiTopic;
+import com.flipkart.varadhi.entities.auth.ResourceType;
 import com.flipkart.varadhi.produce.ProduceResult;
 import com.flipkart.varadhi.produce.config.ProducerOptions;
 import com.flipkart.varadhi.produce.otel.ProducerMetricsEmitter;
 import com.flipkart.varadhi.spi.services.Producer;
-import io.micrometer.core.instrument.MeterRegistry;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
 import lombok.extern.slf4j.Slf4j;
 
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.Objects;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Function;
 
-
+/**
+ * Service responsible for producing messages to topics in Varadhi.
+ * <p>
+ * This service provides high-performance, thread-safe message production capabilities with:
+ * <ul>
+ *   <li>Efficient producer caching using Caffeine for optimal resource utilization</li>
+ *   <li>Fully asynchronous message production with CompletableFuture</li>
+ *   <li>Comprehensive metrics collection for monitoring and performance analysis</li>
+ *   <li>Robust error handling with detailed error messages</li>
+ * </ul>
+ * <p>
+ * The service maintains a cache of producers for storage topics to optimize performance
+ * and resource utilization. Producers are created on-demand and cached for reuse based on
+ * configurable TTL settings.
+ */
 @Slf4j
-public class ProducerService {
-    private final VaradhiCache<StorageTopic, Producer> producerCache;
-    private final VaradhiCache<String, VaradhiTopic> internalTopicCache;
+public final class ProducerService {
+
+    /**
+     * Cache of producers for storage topics.
+     */
+    private final LoadingCache<StorageTopic, Producer> producerCache;
+
+    /**
+     * The region where messages are produced.
+     */
     private final String produceRegion;
 
+    /**
+     * Cache of entity read caches, indexed by resource type.
+     */
+    private final Map<ResourceType, EntityReadCache<?>> readCacheMap;
```

Comment:
why create this hashmap?

We can simply create a field
ReadCache<VaradhiTopic> topicCache;